Copy specs from knative/eventing

tayarani opened the pull request:
<!-- Thanks for sending a pull request! -->

# Changes
As part of https://github.com/knative/specs/issues/4 copying over specs files from knative/eventing
<!-- 
Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! 

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic
-->

-
-
-

<!--
In addition, categorize the changes you're making using the "/kind" Prow command, example:

/kind <kind>

Supported kinds are: api-change, bug, cleanup, deprecation, removal, documentation, enhancement, performance

-->
/kind <kind>

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

<!-- Please include the 'why' behind your changes if no issue exists -->

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->
```release-note

```

**Docs**

<!--
:book: If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [knative/docs]: <issue or pr link>
- [Feature Track]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

tayarani commented:
/assign @vaikas

tayarani commented:
All files from https://github.com/knative/eventing/tree/main/docs/spec moved over as is. Links in data-plane.md were changed to point to new location.

vaikas commented on specs/eventing/README.md:
```diff
@@ -0,0 +1,25 @@
+# Knative Eventing API spec
+
+This directory contains the specification of the Knative Eventing API, which is
+implemented in [`eventing.knative.dev`](/pkg/apis/eventing/v1beta1) and verified
```
since we moved to different repo, we should probably give the full link here and also change to v1 since that's what we're targeting?
https://github.com/knative/eventing/tree/main/pkg/apis/eventing/v1

tayarani replied:
fixed

vaikas commented on specs/eventing/data-plane.md:
```diff
@@ -0,0 +1,125 @@
+If Sink is not returning HTTP success header (200 or 202) then the event may be
+sent again. If the event can not be delivered then some sources of events (such
+as Knative sources, brokers or channels) MAY support
+[dead letter sink or channel](../delivery/README.md) for events that can not be
```
I think this link also needs to be updated.
I'd reckon we'd have a dead-link detection that will flag these and we can follow up since these files need to be vetted with a detailed comb anyways.

tayarani replied:
fixed

vaikas reviewed (approved):
Couple of things:
 1. since it's a move and the docs reference things in the eventing repo, we need to replace some of the refs with full links.
 2. This repo should probably have the invalid link (or whatever robot it is) to flag URLs that don't work. For example, the couple I flagged, there are others.

If this work is just to get a pristine copy with no changes, I'm fine with that and follow up, but I'd think that to ensure that things don't get missed in the future, we should add couple of robots. Also, changing v1beta1 -> v1 should be done soon.

Lastly, I haven't checked yet, but can you put a PR (if you haven't already) in the eventing repo with a "302" to here so that those files are frozen and all the future work happens here.

Thanks for doing this!
/lgtm
/hold

please remove hold at your convenience if the plan above makes sense and we want to checkpoint the work and then start making changes.

tayarani commented:
Thanks for the feedback @vaikas. In response to your comments:
1) I ran a link checker and fixed all the broken link. There were 2 existing broken links which I couldn't fix.
2) I did some searching and looks we don't have a good automated solution for link checking. There was one used in docs but that was disabled and removed. So going forward I will do thing manually until we arrive at a good long term solution.

tayarani commented:
FILE: /tmp/specs/eventing/README.md
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/eventing/v1beta1
[✓] https://github.com/knative/eventing/blob/main/test/e2e
[✓] motivation.md
[✓] overview.md
[✓] interfaces.md
[✓] spec.md
[✓] channel.md
[✓] sources.md
[✓] https://knative.dev/docs/serving/spec/knative-api-specification-1.0/#error-signalling
[✓] https://www.knative.dev/docs/eventing/#architecture

10 links checked.

FILE: /tmp/specs/eventing/broker.md
[✓] ./spec.md#eventingduckv1deliveryspec
[✓] https://github.com/cloudevents/spec/blob/v0.3/http-transport-binding.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md
[✓] https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md
[✓] https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md
[✓] #exceptions

6 links checked.

FILE: /tmp/specs/eventing/channel.md
[✓] https://www.ietf.org/rfc/rfc2119.txt
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1beta1/channelable_types.go
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1/channelable_types.go
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1beta1/subscribable_types.go
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1/subscribable_types.go
[✓] https://github.com/knative/eventing/blob/main/pkg/apis/duck/v1/delivery_types.go
[✓] https://github.com/knative/pkg/blob/main/apis/duck/v1/addressable_types.go
[✓] https://github.com/knative/pkg/blob/main/apis/duck/v1/status_types.go
[✓] https://github.com/cloudevents/spec/blob/v0.3/http-transport-binding.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md#3-http-message-mapping
[✓] #input
[✓] https://w3c.github.io/trace-context/
[✓] https://github.com/knative/eventing/blob/main/config/config-tracing.yaml
[✓] https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md

15 links checked.

FILE: /tmp/specs/eventing/data-plane.md
[✓] ./interfaces.md#addressable
[✓] ./interfaces.md#callable
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md
[✓] https://github.com/cloudevents/spec/blob/v0.3/http-transport-binding.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md#12-relation-to-http
[✓] ./interfaces.md#control-plane
[✓] https://github.com/cloudevents/spec/blob/master/http-protocol-binding.md#3-http-message-mapping
[✓] https://github.com/knative/eventing/blob/main/docs/delivery/README.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/extensions/distributed-tracing.md
[✓] sources.md#source-event-delivery
[✓] channel.md#data-plane
[✓] broker.md

12 links checked.

FILE: /tmp/specs/eventing/helper.md
[✓] spec.md
[✓] #kind-trigger
[✓] #kind-broker
[✓] #kind-channel
[✓] #kind-subscription
[✓] spec.md#kind-trigger
[✓] spec.md#kind-broker
[✓] spec.md#kind-channel
[✓] interfaces.md#addressable
[✓] spec.md#kind-subscription

10 links checked.

FILE: /tmp/specs/eventing/interfaces.md
[✓] motivation.md
[✓] overview.md
[✓] spec.md

3 links checked.

FILE: /tmp/specs/eventing/motivation.md
[✓] https://github.com/cloudevents/spec
[✓] https://github.com/knative/serving
[✓] https://github.com/knative/build
[✓] overview.md
[✓] interfaces.md
[✓] spec.md

6 links checked.

FILE: /tmp/specs/eventing/overview.md
[✓] spec.md#kind-trigger
[✓] spec.md#kind-broker
[✓] spec.md#kind-subscription
[✓] spec.md#kind-channel
[✓] interfaces.md#addressable
[✓] interfaces.md#callable
[✓] interfaces.md
[✓] images/broker-trigger-overview.svg
[✓] images/resource-types-overview.svg
[✓] motivation.md
[✓] spec.md

11 links checked.

FILE: /tmp/specs/eventing/sources.md
[✓] https://cloudevents.io
[✓] ./interfaces.md#addressable
[✓] #source-crds
[✓] 
[✓] #source-custom-objects
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#Source
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#SourceStatus
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/#Condition
[✓] https://github.com/cloudevents/spec/blob/v1.0-rc1/spec.md#type
[✓] https://github.com/cloudevents/spec/blob/v1.0-rc1/spec.md#dataschema
[✓] https://github.com/knative/eventing/blob/main/config/200-source-observer-clusterrole.yaml
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#CloudEventOverrides
[✓] https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#Status
[✓] https://pkg.go.dev/knative.dev/pkg/apis#URL
[✓] https://pkg.go.dev/knative.dev/eventing/pkg/apis/eventing/v1beta1#EventType
[✓] https://knative.dev/docs/eventing/event-registry/
[✓] https://github.com/cloudevents/spec/blob/v0.3/http-transport-binding.md
[✓] https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md
[✓] https://github.com/knative/eventing/blob/main/docs/delivery/README.md

20 links checked.

FILE: /tmp/specs/eventing/spec.md
[✓] interfaces.md
[✓] overview.md
[✓] motivation.md
[✓] #kind-trigger
[✓] #kind-broker
[✓] #kind-channel
[✓] #kind-subscription
[✓] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta
[✓] #triggerfilter
[✓] #duckv1.destination
[✓] #apis.condition
[✓] #apis.url
[✓] #duckv1.kreference
[✓] #eventingduckv1.deliveryspec
[✓] #duckv1.addressable
[✖] duckv1.addressable
[✓] #ChannelTemplateSpec
[✓] #subscriberstatus
[✓] #corev1.ObjectReference
[✓] #duckv1.Destination
[✓] #eventingduckv1.DeliverySpec
[✓] #SubscriptionStatusPhysicalSubscription
[✓] interfaces.md#addressable
[✓] https://golang.org/pkg/net/url/#URL
[✓] https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
[✓] https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
[✖] https://kubernetes.io/docs/concepts/oFURLverview/working-with-objects/names/#names
[✓] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

28 links checked.
[✖] duckv1.addressable → Status: 400
[✖] https://kubernetes.io/docs/concepts/oFURLverview/working-with-objects/names/#names → Status: 404

ERROR: 2 dead links found!

tayarani commented:
On the eventing side I have https://github.com/knative/eventing/pull/5057 ready

tayarani commented:
Ready to merge @vaikas

vaikas commented:
/lgtm
/approve
Thanks!

vaikas commented:
/assign @tayarani

tayarani commented:
/unhold

tayarani commented:
/approve
